File: OpenGSGLibrary/GameDataManager/GameObjectFactory.vb
'    OpenSGSGLibrary is an open-source library for Grand Strategy Games
'    Copyright (C) 2020  Torben Kneesch
'
'    This program is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    This program is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.
Imports System.IO

Namespace WorldData

    ''' <summary>
    ''' Class to generate game objects from game files
    ''' </summary>
    Public Class GameObjectFactory

        ''' <summary>
        ''' Creates a table of generated objects from the parsed files in one directory.
        ''' </summary>
        ''' <typeparam name="baseType">Base variable type of the generated objects (e.g. Province) used for the returned dictionary</typeparam>
        ''' <typeparam name="derivedType">Specialized variable type of the generated objects (e.g. CwpProvince) used for the generated objects.</typeparam>
        ''' <param name="folderPath">String with the directory path.</param>
        ''' <returns>Dictionary with generated objects using file names without extensions as key.</returns>
        Public Shared Function FromFolderWithFilenameId(Of baseType As {New, GameObject}, derivedType As {New, baseType}) _
            (folderPath As String) As Dictionary(Of Integer, baseType)

            If Not Directory.Exists(folderPath) Then
                Throw New DirectoryNotFoundException("Given game data directory not found: " & folderPath)
            End If

            Dim objectTable = New Dictionary(Of Integer, baseType)

            Dim parsedObjectData As Dictionary(Of String, Object) = ParseFolder(folderPath)
            For Each singleObjectData In parsedObjectData
                Dim newObject = New derivedType()
                newObject.SetData(singleObjectData.Key, singleObjectData.Value)
                Dim key As String = ""
                Dim filenameParts As String() = ExtractFromFilename(singleObjectData.Key)
                Try
                    key = Convert.ToInt32(filenameParts(0))
                    objectTable.Add(key, newObject)
                Catch ex As Exception
                    Tools.GlobalLogger.GetInstance().WriteLine(
                        Tools.LogLevel.Fatal, "Error while parsing file: " & singleObjectData.Key)
                End Try
            Next

            Return objectTable

        End Function

        ''' <summary>
        ''' Creates a table of generated objects from the parsed files in one directory.
        ''' </summary>
        ''' <typeparam name="idtype">Variable type of the id (e.g. Integer, String)</typeparam>
        ''' <typeparam name="baseType">Base variable type of the generated objects (e.g. Province) used for the returned dictionary</typeparam>
        ''' <typeparam name="derivedType">Specialized variable type of the generated objects (e.g. CwpProvince) used for the generated objects.</typeparam>
        ''' <param name="folderPath">String with the directory path.</param>
        ''' <param name="keyField">Name of the field in the parsed file which value is used as key in the return value.</param>
        ''' <returns>Dictionary with generated objects.</returns>
        Public Shared Function FromFolder(Of idtype, baseType As {New, GameObject}, derivedType As {New, baseType}) _
            (folderPath As String, keyField As String) As Dictionary(Of idtype, baseType)

            If Not Directory.Exists(folderPath) Then
                Throw New DirectoryNotFoundException("Given game data directory not found: " & folderPath)
            End If

            Dim objectTable = New Dictionary(Of idtype, baseType)

            Dim parsedObjectData As Dictionary(Of String, Object) = ParseFolder(folderPath)
            For Each singleObjectData In parsedObjectData
                Dim countryData As Lookup(Of String, Object) = singleObjectData.Value
                Dim newObject = New derivedType()
                newObject.SetData(singleObjectData.Key, countryData)
                Dim key As idtype = countryData(keyField).Single()
                objectTable.Add(key, newObject)
            Next

            Return objectTable

        End Function

        ''' <summary>
        ''' Creates a list of GameObject objects from a Lookup structure.
        ''' </summary>
        ''' <typeparam name="type">Class inheriting from gamneObject class</typeparam>
        ''' <param name="parsedData">Parser data as a Lookup(Of String, object)</param>
        ''' <param name="objectId">Field id in the top level of the parser data for the 'type' objects</param>
        ''' <returns></returns>
        Public Shared Function ListFromLookup(Of type As {New, GameObject}) _
            (parsedData As Lookup(Of String, Object), objectId As String) As List(Of type)

            Dim generatedList As New List(Of type)

            For Each objParserData In parsedData(objectId)
                Dim newObj As New type()
                newObj.SetData("", objParserData)
                generatedList.Add(newObj)
            Next

            Return generatedList
        End Function

        ''' <summary>
        ''' Parses all files in a folder with txt extension.
        ''' </summary>
        ''' <param name="gamedataPath">String with the path of the folder.</param>
        ''' <returns>Dictionary with a file name as key (without extension) and that file's parsed data as object.</returns>
        Public Shared Function ParseFolder(gamedataPath As String) As Dictionary(Of String, Object)
            Dim dictionaryOfParsedData = New Dictionary(Of String, Object)

            Dim txtFilesInDir As String() = Directory.GetFiles(gamedataPath, "*.txt", SearchOption.AllDirectories)

            Dim scanner = New Parser.Scanner()
            Dim parser = New Parser.Parser()
            For Each textFile As String In txtFilesInDir
                Try
                    Dim rawFile As TextReader = File.OpenText(textFile)
                    Dim nextParseData As Lookup(Of String, Object) = parser.Parse(scanner.Scan(rawFile))
                    If Not IsNothing(nextParseData) Then
                        dictionaryOfParsedData.Add(Path.GetFileNameWithoutExtension(textFile), nextParseData)
                    End If
                Catch ex As Exception
                    'TODO: Need a logger to throw a warning
                End Try
            Next

            Return dictionaryOfParsedData

        End Function

        ''' <summary>
        ''' Returns parts of a file name separated by - without path and extension 
        ''' </summary>
        ''' <param name="filePath">File name with extension and possibly path.</param>
        ''' <returns>Array of strings with the parts of the file names.</returns>
        Public Shared Function ExtractFromFilename(filePath As String) As String()
            Dim filename As String = Path.GetFileNameWithoutExtension(filePath)
            If filename = "" Then
                Throw New ApplicationException("No file name found in path: " + filePath)
            End If

            Dim nameParts As String() = filename.Split("-")
            For i = 0 To UBound(nameParts)
                nameParts(i) = nameParts(i).Trim()
            Next

            Return nameParts
        End Function

    End Class

End Namespace

File: OpenGSGLibrary2/GameDataManager/GameObjectFactory.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;

namespace WorldData
{
    /// <summary>
    /// Class to generate game objects from game files
    /// </summary>
    public static class GameObjectFactory
    {
        /// <summary>
        /// Creates a table of generated objects from the parsed files in one directory.
        /// </summary>
        /// <typeparam name="baseType">Base variable type of the generated objects (e.g. Province) used for the returned dictionary</typeparam>
        /// <typeparam name="derivedType">Specialized variable type of the generated objects (e.g. CwpProvince) used for the generated objects.</typeparam>
        /// <param name="folderPath">String with the directory path.</param>
        /// <returns>Dictionary with generated objects using file names without extensions as key.</returns>
        public static Dictionary<int, TBase> FromFolderWithFilenameId<TBase, TDerived>(
            string folderPath
        )
            where TBase : GameObject, new()
            where TDerived : TBase, new()
        {
            if (!Directory.Exists(folderPath))
                throw new DirectoryNotFoundException(
                    "Given game data directory not found: " + folderPath
                );

            var objectTable = new Dictionary<int, TBase>();

            var parsedObjectData = ParseFolder(folderPath);
            foreach (var singleObjectData in parsedObjectData)
            {
                var newObject = new TDerived();
                newObject.SetData(
                    singleObjectData.Key,
                    (ILookup<string, object>)singleObjectData.Value
                );

                try
                {
                    var filenameParts = ExtractFromFilename(singleObjectData.Key);
                    var key = Convert.ToInt32(filenameParts[0]);
                    objectTable.Add(key, newObject);
                }
                catch (Exception)
                {
                    // Ignore entries that cannot be parsed; original code logged a fatal error.
                }
            }

            return objectTable;
        }

        /// <summary>
        /// Creates a table of generated objects from the parsed files in one directory.
        /// </summary>
        /// <typeparam name="idtype">Variable type of the id (e.g. Integer, String)</typeparam>
        /// <typeparam name="baseType">Base variable type of the generated objects (e.g. Province) used for the returned dictionary</typeparam>
        /// <typeparam name="derivedType">Specialized variable type of the generated objects (e.g. CwpProvince) used for the generated objects.</typeparam>
        /// <param name="folderPath">String with the directory path.</param>
        /// <param name="keyField">Name of the field in the parsed file which value is used as key in the return value.</param>
        /// <returns>Dictionary with generated objects.</returns>
        public static Dictionary<TKey, TBase> FromFolder<TKey, TBase, TDerived>(
            string folderPath,
            string keyField
        )
            where TBase : GameObject, new()
            where TDerived : TBase, new()
            where TKey : notnull
        {
            if (!Directory.Exists(folderPath))
                throw new DirectoryNotFoundException(
                    "Given game data directory not found: " + folderPath
                );

            var objectTable = new Dictionary<TKey, TBase>();

            var parsedObjectData = ParseFolder(folderPath);
            foreach (var singleObjectData in parsedObjectData)
            {
                var countryData = (ILookup<string, object>)singleObjectData.Value;
                var newObject = new TDerived();
                newObject.SetData(singleObjectData.Key, countryData);

                var keyObj = countryData[keyField].Single();
                TKey key = (TKey)Convert.ChangeType(keyObj, typeof(TKey));
                objectTable.Add(key, newObject);
            }

            return objectTable;
        }

        /// <summary>
        /// Creates a list of GameObject objects from a Lookup structure.
        /// </summary>
        /// <typeparam name="type">Class inheriting from gamneObject class</typeparam>
        /// <param name="parsedData">Parser data as a Lookup(Of String, object)</param>
        /// <param name="objectId">Field id in the top level of the parser data for the 'type' objects</param>
        /// <returns></returns>
        public static List<T> ListFromLookup<T>(ILookup<string, object> parsedData, string objectId)
            where T : GameObject, new()
        {
            var generatedList = new List<T>();

            foreach (var objParserData in parsedData[objectId])
            {
                var newObj = new T();
                newObj.SetData(string.Empty, (ILookup<string, object>)objParserData);
                generatedList.Add(newObj);
            }

            return generatedList;
        }

        /// <summary>
        /// Parses all files in a folder with txt extension.
        /// </summary>
        /// <param name="gamedataPath">String with the path of the folder.</param>
        /// <returns>Dictionary with a file name as key (without extension) and that file's parsed data as object.</returns>
        public static Dictionary<string, object> ParseFolder(string gamedataPath)
        {
            var dictionaryOfParsedData = new Dictionary<string, object>();

            var txtFilesInDir = Directory.GetFiles(
                gamedataPath,
                "*.txt",
                SearchOption.AllDirectories
            );

            foreach (var textFile in txtFilesInDir)
            {
                try
                {
                    using var rawFile = File.OpenText(textFile);

                    // Use the Parser.Scanner and Parser.Parser types which are included in the project
                    var scanner = new Parser.Scanner();
                    var parser = new Parser.Parser();

                    var tokenStream = scanner.Scan(rawFile);
                    var nextParseData = parser.Parse(tokenStream);
                    if (nextParseData != null)
                    {
                        dictionaryOfParsedData.Add(
                            Path.GetFileNameWithoutExtension(textFile),
                            nextParseData
                        );
                    }
                }
                catch (Exception)
                {
                    // ignore parse errors
                }
            }

            return dictionaryOfParsedData;
        }

        /// <summary>
        /// Returns parts of a file name separated by - without path and extension
        /// </summary>
        /// <param name="filePath">File name with extension and possibly path.</param>
        /// <returns>Array of strings with the parts of the file names.</returns>
        public static string[] ExtractFromFilename(string filePath)
        {
            var filename = Path.GetFileNameWithoutExtension(filePath);
            if (string.IsNullOrEmpty(filename))
                throw new ApplicationException("No file name found in path: " + filePath);

            var nameParts = filename.Split('-', StringSplitOptions.RemoveEmptyEntries);
            for (int i = 0; i < nameParts.Length; i++)
                nameParts[i] = nameParts[i].Trim();

            return nameParts;
        }
    }
}
